File: android/app/src/main/java/sol/vortexx/android/chat/impl/ChatEngine.kt
package sol.vortexx.android.chat.impl

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import sol.vortexx.android.chat.api.IncomingMessages
import sol.vortexx.android.chat.api.MessageSender
import sol.vortexx.android.chat.api.SendOutcome
import sol.vortexx.android.crypto.api.Aead
import sol.vortexx.android.crypto.util.Hex
import sol.vortexx.android.db.dao.MessageDao
import sol.vortexx.android.db.entities.MessageEntity
import sol.vortexx.android.keys.api.KeyAcquisition
import sol.vortexx.android.keys.api.RoomKeyProvider
import sol.vortexx.android.search.api.SearchRepository
import sol.vortexx.android.ws.api.WsClient
import sol.vortexx.android.ws.api.WsFrame
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.random.Random

/**
 * End-to-end chat engine. Owns three responsibilities:
 *   1. Outbound encrypt + persist + WS send
 *   2. Inbound WS receive → decrypt → persist
 *   3. On key-arrival events, re-decrypt any messages that previously
 *      landed while the room key was still Pending.
 *
 * Every plaintext that lands here is also pushed into [SearchRepository]
 * so full-text search covers every delivered message.
 */
@Singleton
class ChatEngine @Inject constructor(
    private val ws: WsClient,
    private val aead: Aead,
    private val keys: RoomKeyProvider,
    private val dao: MessageDao,
    private val search: SearchRepository,
) : MessageSender, IncomingMessages {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private val json  = Json { ignoreUnknownKeys = true }

    init { scope.launch { observeIncoming() } }

    override fun messagesIn(roomId: Long): Flow<List<MessageEntity>> = dao.observeRoom(roomId)

    override suspend fun send(roomId: Long, plaintext: String): SendOutcome {
        val key = when (val r = keys.keyFor(roomId)) {
            is KeyAcquisition.Ready   -> r.keyHex
            is KeyAcquisition.Pending -> return SendOutcome.Error("key_pending:${r.reason}")
            is KeyAcquisition.Error   -> return SendOutcome.Error(r.reason)
        }
        val packed = aead.encrypt(Hex.decode(key), plaintext.toByteArray(Charsets.UTF_8))
        val ctHex  = Hex.encode(packed)

        val localId = Random.nextLong(Long.MAX_VALUE - 1) + 1
        val msg = MessageEntity(
            id = localId, roomId = roomId, plaintext = plaintext,
            ciphertextHex = ctHex, sentAt = System.currentTimeMillis(),
        )
        dao.upsert(msg)
        search.index(msg.id, plaintext)
        ws.send(WsFrame(json.encodeToString(
            OutFrame.serializer(),
            OutFrame(action = "send_message", room_id = roomId, ciphertext = ctHex),
        )))
        return SendOutcome.Queued(localId)
    }

    /**
     * WS receive loop.
     *   - `peer_message` → decrypt + store + index
     *   - `public_room_key_updated` / `room_key` → re-decrypt any backlog
     *     that was stored with plaintext=null because the key wasn't ready.
     */
    private suspend fun observeIncoming() {
        ws.incoming.collect { frame ->
            val obj = runCatching { json.parseToJsonElement(frame.text).jsonObject }.getOrNull()
                ?: return@collect
            when (obj["type"]?.jsonPrimitive?.content) {
                "peer_message" -> onPeerMessage(obj)
                "public_room_key_updated", "room_key", "room_key_rotated" -> {
                    val roomId = obj["room_id"]?.jsonPrimitive?.content?.toLongOrNull() ?: return@collect
                    retryUndecrypted(roomId)
                }
                else -> { /* not ours */ }
            }
        }
    }

    private suspend fun onPeerMessage(obj: kotlinx.serialization.json.JsonObject) {
        val roomId = obj["room_id"]?.jsonPrimitive?.content?.toLongOrNull() ?: return
        val ctHex  = obj["ciphertext"]?.jsonPrimitive?.content ?: return
        val sender = obj["sender"]?.jsonPrimitive?.content
        val id     = obj["msg_id"]?.jsonPrimitive?.content?.toLongOrNull() ?: System.nanoTime()

        val plaintext = tryDecrypt(roomId, ctHex)
        val msg = MessageEntity(
            id = id, roomId = roomId, senderUsername = sender,
            plaintext = plaintext, ciphertextHex = ctHex,
            sentAt = System.currentTimeMillis(),
        )
        dao.upsert(msg)
        if (plaintext != null) search.index(id, plaintext)
    }

    /**
     * Walk the backlog of not-yet-decrypted messages in [roomId] and try
     * to decrypt them now that a fresh key has arrived. Idempotent —
     * already-decrypted rows are left alone by [MessageDao.undecrypted].
     */
    private suspend fun retryUndecrypted(roomId: Long) {
        val pending = dao.undecrypted(roomId)
        if (pending.isEmpty()) return
        for (row in pending) {
            val plaintext = tryDecrypt(roomId, row.ciphertextHex) ?: continue
            dao.upsert(row.copy(plaintext = plaintext))
            search.index(row.id, plaintext)
        }
    }

    private suspend fun tryDecrypt(roomId: Long, ctHex: String): String? {
        val keyAcq = keys.keyFor(roomId)
        if (keyAcq !is KeyAcquisition.Ready) return null
        return runCatching {
            String(aead.decrypt(Hex.decode(keyAcq.keyHex), Hex.decode(ctHex)), Charsets.UTF_8)
        }.getOrNull()
    }

    @Serializable
    private data class OutFrame(val action: String, val room_id: Long, val ciphertext: String)
}

File: ios/Modules/Sources/Chat/impl/ChatEngine.swift
import Foundation
import VortexCrypto
import DB
import Keys
import WS

/// End-to-end chat engine.
///
/// Mirrors the Android `ChatEngine`:
///   * send → AEAD encrypt with room key → WS frame + local DB row
///   * receive → decode JSON → decrypt → DB upsert
///   * on `public_room_key_updated` / `room_key` re-decrypts stored
///     messages that were previously pending a key.
public final class ChatEngine: MessageSender, IncomingMessages, @unchecked Sendable {
    private let ws: WsClient
    private let aead: Aead
    private let keys: RoomKeyProvider
    private let dao: MessageDao
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()
    private let runTask: Task<Void, Never>

    public init(ws: WsClient, aead: Aead, keys: RoomKeyProvider, dao: MessageDao) {
        self.ws = ws
        self.aead = aead
        self.keys = keys
        self.dao = dao
        // Must capture the stream *before* calling `start()` so we don't
        // race with reconnect emitting the first frame.
        let incoming = ws.incoming
        self.runTask = Task { [aead, keys, dao] in
            await ChatEngine.observeIncoming(incoming: incoming, aead: aead, keys: keys, dao: dao)
        }
    }

    public func messagesIn(_ roomId: Int64) -> AsyncStream<[MessageRecord]> {
        dao.observeRoom(roomId)
    }

    public func send(roomId: Int64, plaintext: String) async -> SendOutcome {
        let keyHex: String
        switch await keys.keyFor(roomId) {
        case .ready(let hex, _): keyHex = hex
        case .pending(let r):    return .error("key_pending:\(r)")
        case .error(let r):      return .error(r)
        }
        do {
            let key = try Hex.decode(keyHex)
            let packed = try aead.encrypt(key: key, plaintext: Data(plaintext.utf8))
            let ctHex = Hex.encode(packed)
            let localId = Int64.random(in: 1...Int64.max)
            try await dao.upsert(MessageRecord(
                id: localId, roomId: roomId,
                plaintext: plaintext, ciphertextHex: ctHex,
                sentAt: Int64(Date().timeIntervalSince1970 * 1000),
            ))
            let frame = OutFrame(action: "send_message", room_id: roomId, ciphertext: ctHex)
            if let json = try? encoder.encode(frame), let s = String(data: json, encoding: .utf8) {
                await ws.send(s)
            }
            return .queued(localId: localId)
        } catch {
            return .error((error as NSError).localizedDescription)
        }
    }

    // ── incoming loop ──────────────────────────────────────────────────

    private static func observeIncoming(
        incoming: AsyncStream<String>,
        aead: Aead,
        keys: RoomKeyProvider,
        dao: MessageDao,
    ) async {
        let decoder = JSONDecoder()
        for await text in incoming {
            guard let data = text.data(using: .utf8) else { continue }
            guard let generic = try? decoder.decode(GenericFrame.self, from: data) else { continue }
            switch generic.type {
            case "peer_message":
                await onPeerMessage(data, aead: aead, keys: keys, dao: dao)
            case "public_room_key_updated", "room_key", "room_key_rotated":
                if let rid = generic.room_id { await retryUndecrypted(rid, aead: aead, keys: keys, dao: dao) }
            default: break
            }
        }
    }

    private static func onPeerMessage(
        _ data: Data, aead: Aead, keys: RoomKeyProvider, dao: MessageDao,
    ) async {
        guard let m = try? JSONDecoder().decode(PeerMessage.self, from: data) else { return }
        let plaintext = await tryDecrypt(roomId: m.room_id, ctHex: m.ciphertext, aead: aead, keys: keys)
        let row = MessageRecord(
            id: m.msg_id ?? Int64(Date().timeIntervalSince1970 * 1_000_000),
            roomId: m.room_id,
            senderUsername: m.sender,
            plaintext: plaintext,
            ciphertextHex: m.ciphertext,
            sentAt: Int64(Date().timeIntervalSince1970 * 1000),
        )
        try? await dao.upsert(row)
    }

    private static func retryUndecrypted(
        _ roomId: Int64, aead: Aead, keys: RoomKeyProvider, dao: MessageDao,
    ) async {
        guard let pending = try? await dao.undecrypted(roomId), !pending.isEmpty else { return }
        for var row in pending {
            guard let plaintext = await tryDecrypt(
                roomId: roomId, ctHex: row.ciphertextHex, aead: aead, keys: keys,
            ) else { continue }
            row.plaintext = plaintext
            try? await dao.upsert(row)
        }
    }

    private static func tryDecrypt(
        roomId: Int64, ctHex: String, aead: Aead, keys: RoomKeyProvider,
    ) async -> String? {
        guard case .ready(let keyHex, _) = await keys.keyFor(roomId) else { return nil }
        guard let key = try? Hex.decode(keyHex),
              let packed = try? Hex.decode(ctHex),
              let plain = try? aead.decrypt(key: key, packed: packed)
        else { return nil }
        return String(data: plain, encoding: .utf8)
    }

    // ── DTOs ───────────────────────────────────────────────────────────

    private struct OutFrame: Codable {
        let action: String
        let room_id: Int64
        let ciphertext: String
    }
    private struct GenericFrame: Decodable {
        let type: String?
        let room_id: Int64?
    }
    private struct PeerMessage: Decodable {
        let room_id: Int64
        let ciphertext: String
        let sender: String?
        let msg_id: Int64?
    }
}
